include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+            using type1 = xtl::mpl::eval_if_t<std::is_fundamental<decay_t1>, id<decay_t1>, get_value_type<decay_t1>>;
+            using type2 = xtl::mpl::eval_if_t<std::is_fundamental<decay_t2>, id<decay_t2>, get_value_type<decay_t2>>;
+            using type = xmasked_value<std::common_type_t<type1, type2>>;
+        };
+
+        template <class T1, class T2, class B2>
+        struct common_masked_value_impl<T1, xmasked_value<T2, B2>>
+            : common_masked_value_impl<T1, T2>
+        {
+        };
+
+        template <class T1, class B1, class T2>
+        struct common_masked_value_impl<xmasked_value<T1, B1>, T2>
+            : common_masked_value_impl<T2, xmasked_value<T1, B1>>
+        {
+        };
+
+        template <class T1, class B1, class T2, class B2>
+        struct common_masked_value_impl<xmasked_value<T1, B1>, xmasked_value<T2, B2>>
+            : common_masked_value_impl<T1, T2>
+        {
+        };
+
+        template <class T1, class T2, class... Args>
+        struct common_masked_value_impl<T1, T2, Args...>
+        {
+            using type = typename common_masked_value_impl<
+                             typename common_masked_value_impl<T1, T2>::type,
+                             Args...
+                         >::type;
+        };
+    }
+
+    template <class E>
+    using is_xmasked_value = detail::is_xmasked_value_impl<E>;
+
+    template <class E, class R>
+    using disable_xoptional_like = std::enable_if_t<!xtl::is_xoptional<E>::value && !is_xmasked_value<E>::value, R>;
+
+    template <class E1, class E2, class R = void>
+    using disable_both_xoptional_like = std::enable_if_t<
+        !xtl::is_xoptional<E1>::value && !is_xmasked_value<E1>::value &&
+            !xtl::is_xoptional<E2>::value && !is_xmasked_value<E2>::value,
+        R
+    >;
+
+    template <class... Args>
+    struct common_masked_value : detail::common_masked_value_impl<Args...>
+    {
+    };
+
+    template <class... Args>
+    using common_masked_value_t = typename common_masked_value<Args...>::type;
+
+    /****************
+    * xmasked_value *
+    *****************/
+
+    template <class T, class B>
+    class xmasked_value
+    {
+    public:
+
+        using self_type = xmasked_value<T, B>;
+
+        using value_type = T;
+        using flag_type = B;
+        using optional_type = xtl::xoptional<value_type, flag_type>;
+
+        template <class T1, class B1>
+        inline constexpr xmasked_value(T1&& value, B1&& flag)
+            : m_value(std::forward<T1>(value)), m_flag(std::forward<B1>(flag))
+        {
+        }
+
+        inline explicit constexpr xmasked_value(const xtl::xoptional<T, B>& opt)
+            : m_value(opt.value()), m_flag(opt.has_value())
+        {
+        }
+
+        inline operator value_type() {
+            return m_value;
+        }
+
+        inline operator optional_type() {
+            return optional_type(m_value, m_flag);
+        }
```

Comment:
There should be only one conversion operator to `value_type` (which should be `const`), which handles both cases (i.e. `value_type` is `xoptional` and `value_type` is not). A way to implement that is to forward the call to implementation methods enabled / disabled with SFINAE:

```cpp
template <class V>
std::enable_if_t<xtl::is_xoptional<V>::value, V> convert() const
{
    return xoptional(value().value(), value().has_value() && has_value());
}
template <class V>
std::enable_if_t<!xtl::is_xoptional<V>::value, V> convert() const
{
    return value();
}
```